man/hfst-fst2strings.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HFST-FST2STRINGS "1" "April 2026" "HFST" "User Commands"
.SH NAME
hfst-fst2strings \- =Display the strings recognized by a transducer
.SH SYNOPSIS
.B hfst-fst2strings
[\fI\,OPTIONS\/\fR...] [\fI\,INFILE\/\fR]
.SH DESCRIPTION
Display the strings recognized by a transducer
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-\-colour\fR[=\fI\,WHEN\/\fR]
Print in colour WHEN:
.TP
\fB\-\-color\fR[=\fI\,WHEN\/\fR]
always, never, auto (default)
.SS "Fst2strings options:"
.TP
\fB\-n\fR, \fB\-\-max\-strings\fR=\fI\,NSTR\/\fR
print at most NSTR strings
.TP
\fB\-N\fR, \fB\-\-nbest\fR=\fI\,NBEST\/\fR
print at most NBEST best strings
.TP
\fB\-r\fR, \fB\-\-random\fR=\fI\,NRAND\/\fR
print at most NRAND random strings
.TP
\fB\-c\fR, \fB\-\-cycles\fR=\fI\,NCYC\/\fR
follow cycles at most NCYC times
.TP
\fB\-w\fR, \fB\-\-print\-weights\fR
display the weight for each string
.TP
\fB\-S\fR, \fB\-\-print\-separator\fR
print separator "\-\-" after each transducer
.TP
\fB\-e\fR, \fB\-\-epsilon\-format\fR=\fI\,EPS\/\fR
print epsilon as EPS
.TP
\fB\-X\fR, \fB\-\-xfst\fR=\fI\,VARIABLE\/\fR
toggle xfst compatibility option VARIABLE
.SS "Path filters:"
.TP
\fB\-b\fR, \fB\-\-beam\fR=\fI\,B\/\fR
reject output string with weight more than B away from
the weight of the best output string
.TP
\fB\-l\fR, \fB\-\-max\-in\-length\fR=\fI\,MIL\/\fR
reject input string longer than MIL
.TP
\fB\-L\fR, \fB\-\-max\-out\-length\fR=\fI\,MOL\/\fR
reject output string longer than MOL
.TP
\fB\-p\fR, \fB\-\-in\-prefix\fR=\fI\,OPREFIX\/\fR
input string must begin with IPREFIX
.TP
\fB\-P\fR, \fB\-\-out\-prefix\fR=\fI\,OPREFIX\/\fR
output string must begin with OPREFIX
.TP
\fB\-u\fR, \fB\-\-in\-exclude\fR=\fI\,IXSTR\/\fR
input string must not contain IXSTR
.TP
\fB\-U\fR, \fB\-\-out\-exclude\fR=\fI\,OXST\/\fR
output string must not contain OXSTR
.PP
If OUTFILE or INFILE is missing or \-, standard streams will be used.
Format of result depends on format of INFILE
If all NSTR, NBEST and NCYC are omitted, all possible paths are printed:
NSTR, NBEST and NCYC default to infinity.
NBEST overrides NSTR and NCYC
NRAND overrides NBEST, NSTR and NCYC
B must be a non\-negative float
If EPS is not given, default is empty string.
Numeric options are parsed with strtod(3).
Xfst variables supported are { obey\-flags, print\-flags,
print\-pairs, print\-space, quote\-special }.
.SH EXAMPLES
.TP
hfst\-fst2strings lexical.hfst
generates all forms of lexical.hfst
.TP
hfst\-fst2strings \-P "cat<n>" \-c 0 lexical.hfst
generates paradigm for cat<n> without following cycles
.SS "Known bugs:"
.IP
Does not work correctly for hfst optimized lookup format.
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-fst2strings home page:
<https://github.com/hfst/hfst/wiki/HfstFst2Strings>
.br
General help using HFST software:
<https://github.com/hfst/hfst/wiki>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS for hfst-tools and library

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

## Noteworthy changes in 3.17.1

* fixes logic error where `hfst-lexc` would treat a warning about implicit
  alphabet in XREs as a critical error even when `-Werror` is not asked

## Noteworthy changes in 3.17.0

* API breaks: HfstTokenizer, LexcCompiler, `read_in_att_format` functions
  in HfstTransducer, HfstBasicTransducer, etc. to accommodate the following
  unicode handling changes and error reporting:
* Character "tokenisers" that used to operate on bytes or UTF-8 sequences will
  now operate on [graphemes](https://www.unicode.org/reports/tr29/tr29-23.html#Grapheme_Cluster_Boundaries),
  this means that by default arcs in FSAs are expected to match combining
  grapheme clusters, e.g. a base character plus all diacritics.
  The implementation is based on ICU implementation of `u_brk`
    * in lexc string entries, graphemes are compiled into a arcs
    * in xerox regex curly braces, as well
    * in tokenise without --multichar mode, input is tokenised by graphemens
    * etc. etc. tbd
* negative weights yield warnings now unless turned off
* tools will have more gcc-style warning control: -Wall -Werror -Wno-this-error
* lexc warnings are systematically treated as errors with -Werror now
* colourised more warnings and errors

## Noteworthy changes in 3.16.2

* Support for `>OpenFst-1.8.0`
    * only functional change is removal of n-best paths search for log weight
      automata, it probably has errored out before but has slightly different
      error now (HfstException instead of openfst exit)
* Optimisations to hfst-pmatch2fst alphabet handling

## Noteworthy changes in 3.16.1

* Bug fixes:

    * hfst-tokenize escapes both \r and \n now

    * regexp parsing more robust with comments and linebreaks

* Removal of some bundled libraries including foma and openfst, install
  suitable versions to compile support for them

* ICU stabilised as unicode handler over all

* experimental C API for rust bindings (unstable)

* new tools (testing):

    * hfst-check-alpha for checking and comparing alphabets in automata and
      their metadata

    * hfst-dump-alphabets for reviewing alphabets easier than hfst-summarize
      grepping

    * hfst-kill-paths for efficient path removal

* introduced clang-format to work towards standard code formatting

Noteworthy changes in 3.16.0
----------------------------

* Bugfixes:

  * During conversion between optimized-lookup and other formats, the alphabet
    was not being preserved, leading to incorrect behaviour in some cases.
    This has been fixed.

  * Fix segfaults in hfst-summarize, hfst-disjunct and foma back-end.

  * Force ICU, fixes #501.

  * Use unique_ptr for input/ouput stream memory allocation (#547).


* Improvements:

  * hfst-pmatch archives now no longer include or parse redundant alphabets,
    which in some cases can be a meaningful size and speed saving.

  * Multiple memory leaks fixed.

  * Drop support for TR1 standard assuming that C++14 is minimal requirement.

  * Drop legacy wrappers for foma and lexc (#532).

  * Use upstream Foma and by default upstream OpenFST.

  * Use C++17, or C++14 with bundled OpenFST.

  * Update external links for sfst, foma and flex.

  * Add lex/yacc sources to dist (see issue #524).

* Features:

  * A new variable in the hfst-pmatch language, "unicode-character-classes",
    which causes the use of built-in character sets to be interpreted at
    runtime using a unicode library.

  * Verbose output from hfst-pmatch2fst has been much improved, providing eg.
    a call graph of the process. It also warns about use of pmatch features
    that is likely to be incorrect and lead to errors.


Noteworthy changes in 3.15.5
----------------------------

* Fixes to pmatch:

  * Modify tracking of entry points to allow for RTNs in Captures()

  * Fix off-by-one bug in checking captures

* Fixes to other issues: #508, #488, #486, #452, #491, #494

* Fix memory leaks


Noteworthy changes in 3.15.4
----------------------------

* Import Foma patches with license change to Apache-2.0 and minimal other fixes

* Allow regeneration of Foma files so the generated files can be omitted

* Explicit license file for SFST


Noteworthy changes in 3.15.3
----------------------------

* Numerous improvements to command line tools, e.g. substitute, insert-freely, proc, fst2txt, expand


Noteworthy changes in 3.15.2
----------------------------

* Fix off-by-one bug in checking captures in pmatch


Noteworthy changes in 3.15.1
----------------------------

* Fix issue #443

* Fix issue #429

* Fix issue #409


Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize


tools/src/hfst-fst2strings.cc
//! @file hfst-fst2strings.cc
//!
//! @brief Transducer path printing command line tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#ifdef WINDOWS
#include <io.h>
#endif


#include <iostream>
#include <fstream>
#include <memory>

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "HfstTransducer.h"
#include "HfstInputStream.h"
#include "HfstOutputStream.h"
#include "HfstSymbolDefs.h"

using hfst::HfstTransducer;
using hfst::HfstInputStream;
using hfst::FdOperation;

//using hfst::WeightedPaths;
//using hfst::WeightedPath;
using hfst::StringPairVector;
using hfst::HfstTwoLevelPath;
using hfst::HfstTwoLevelPaths;


using hfst::HFST_OL_TYPE;
using hfst::HFST_OLW_TYPE;

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

// the maximum number of strings printed for each transducer
static int max_strings = 0;
static int cycles = -1;
static int nbest_strings=-1;
static int max_random_strings=-1;
static float max_weight=-1;
static float beam=-1;
static bool display_weights=false;
static bool eval_fd=false;
static bool filter_fd=true;
static bool quote_special=false;
static bool print_spaces=false;
static unsigned int max_input_length = 0;
static unsigned int max_output_length = 0;
static std::string input_prefix;
static std::string output_prefix;
static std::string input_exclude;
static std::string output_exclude;

static bool print_in_pairstring_format=false;
static char * epsilon_format=0;

static bool print_separator_after_each_transducer=false;

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
        "Display the strings recognized by a transducer\n"
        "\n", program_name);
    print_common_program_options(message_out);
    fprintf(message_out, "Fst2strings options:\n"
"  -n, --max-strings=NSTR     print at most NSTR strings\n"
"  -N, --nbest=NBEST          print at most NBEST best strings\n"
"  -r, --random=NRAND         print at most NRAND random strings\n"
"  -c, --cycles=NCYC          follow cycles at most NCYC times\n"
"  -w, --print-weights        display the weight for each string\n"
"  -S, --print-separator      print separator \"--\" after each transducer\n"
"  -e, --epsilon-format=EPS   print epsilon as EPS\n"
"  -X, --xfst=VARIABLE        toggle xfst compatibility option VARIABLE\n");
    fprintf(message_out, "Path filters:\n"
"  -b, --beam=B               reject output string with weight more than B away from\n"
"                             the weight of the best output string\n"
"  -l, --max-in-length=MIL    reject input string longer than MIL\n"
"  -L, --max-out-length=MOL   reject output string longer than MOL\n"
"  -p, --in-prefix=OPREFIX    input string must begin with IPREFIX\n"
"  -P, --out-prefix=OPREFIX   output string must begin with OPREFIX\n"
"  -u, --in-exclude=IXSTR     input string must not contain IXSTR\n"
"  -U, --out-exclude=OXST     output string must not contain OXSTR\n");

    fprintf(message_out, "\n");

    print_common_unary_program_parameter_instructions(message_out);
    fprintf(message_out, "If all NSTR, NBEST and NCYC are omitted, "
            "all possible paths are printed:\n"
            "NSTR, NBEST and NCYC default to infinity.\n"
            "NBEST overrides NSTR and NCYC\n"
        "NRAND overrides NBEST, NSTR and NCYC\n"
            "B must be a non-negative float\n"
            "If EPS is not given, default is empty string.\n"
            "Numeric options are parsed with strtod(3).\n"
        "Xfst variables supported are { obey-flags, print-flags,\n"
        "print-pairs, print-space, quote-special }.\n");
    fprintf(message_out,
        "\n"
        "Examples:\n"
        "  %s lexical.hfst    generates all forms of lexical.hfst\n"
        "  %s -P \"cat<n>\" -c 0 lexical.hfst\n"
        "                     generates paradigm for cat<n> without following cycles\n"
        "\n", program_name, program_name);

    fprintf(message_out,
        "Known bugs:\n"
        "  Does not work correctly for hfst optimized lookup format.\n"
        "\n"
        );

    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
}

int
parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
          {
            HFST_GETOPT_COMMON_LONG,
            HFST_GETOPT_UNARY_LONG,
            {"beam", required_argument, 0, 'b'},
            {"cycles", required_argument, 0, 'c'},
            {"epsilon-format", required_argument, 0, 'e'},
            {"in-exclude", required_argument, 0, 'u'},
            {"in-prefix", required_argument, 0, 'p'},
            {"max-in-length", required_argument, 0, 'l'},
            {"max-out-length", required_argument, 0, 'L'},
            {"max-strings", required_argument, 0, 'n'},
            {"nbest", required_argument, 0, 'N'},
            {"random", required_argument, 0, 'r'},
            {"print-separator", no_argument, 0, 'S'},
            {"out-exclude", required_argument, 0, 'U'},
            {"out-prefix", required_argument, 0, 'P'},
            {"print-weights", no_argument, 0, 'w'},
            {"xfst", required_argument, 0, 'X'},
            {0,0,0,0}
          };
        int option_index = 0;
        int c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             HFST_GETOPT_UNARY_SHORT
                             "Swb:c:e:u:p:l:L:n:r:N:U:P:X:",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }
        //char *level = NULL;
        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
        case 'n':
            max_strings = hfst_strtoul(optarg, 10);
            break;
        case 'N':
            nbest_strings = hfst_strtoul(optarg, 10);
            break;
        case 'r':
            max_random_strings = hfst_strtoul(optarg, 10);
            break;
        case 'b':
          beam = atof(optarg);
          if (beam < 0)
            {
              std::cerr << "Invalid argument for --beam\n";
              return EXIT_FAILURE;
            }
          break;
        case 'c':
            cycles = hfst_strtoul(optarg, 10);
            break;
        case 'w':
            display_weights = true;
            break;
        case 'X':
            if (strcmp(optarg, "obey-flags") == 0)
              {
                eval_fd = true;
              }
            else if (strcmp(optarg, "print-flags") == 0)
              {
                filter_fd = false;
              }
            else if (strcmp(optarg, "quote-special") == 0)
              {
                quote_special = true;
              }
            else if (strcmp(optarg, "print-pairs") == 0)
              {
                print_in_pairstring_format = true;
              }
            else if (strcmp(optarg, "print-space") == 0)
              {
                print_spaces = true;
              }
            else
              {
                error(0, EXIT_FAILURE, "Unrecognised xfst option. "
                     "available options are obey-flags, print-flags\n");
              }
          break;
        case 'l':
            max_input_length = hfst_strtoul(optarg, 10);
            break;
        case 'L':
            max_output_length = hfst_strtoul(optarg, 10);
            break;
        case 'p':
            input_prefix = optarg;
            break;
        case 'P':
            output_prefix = optarg;
            break;
        case 'u':
            input_exclude = optarg;
            break;
        case 'U':
          output_exclude = optarg;
          break;
        case 'S':
          print_separator_after_each_transducer = true;
          break;
        case 'e':
          epsilon_format = hfst_strdup(optarg);
          break;
#include "inc/getopt-cases-error.h"
        }
    }

#include "inc/check-params-common.h"
#include "inc/check-params-unary.h"
    return EXIT_CONTINUE;
}

/* Replace all strings \a str1 in \a symbol with \a str2. */
static std::string replace_all(std::string symbol,
                   const std::string &str1,
                   const std::string &str2)
{
  size_t pos = symbol.find(str1);
  while (pos != std::string::npos) // while there are str1:s to replace
    {
      symbol.erase(pos, str1.size()); // erase str1
      symbol.insert(pos, str2);       // insert str2 instead
      pos = symbol.find               // find next str1
    (str1, pos+str2.size());
    }
  return symbol;
}


static std::string get_print_format(const std::string &s)
{
  // print epsilon as defined by the user or use the default
  if (hfst::is_epsilon(s))
    return std::string(strdup(epsilon_format));

  if (not quote_special)
    return std::string(s);

  // escape spaces and colons as they have a special meaning
  return
    replace_all
    (replace_all
     ( replace_all
       (std::string(s), " ", "@_SPACE_@"),
       ":", "@_COLON_@"),
     "\t", "@_TAB_@");
}

//Print results as they come
class Callback : public hfst::ExtractStringsCb
{
public:
  int count;
  int max_num;
  std::ostream* out_;
  Callback(int max, std::ostream* out): count(0), max_num(max), out_(out) {}
  RetVal operator()(HfstTwoLevelPath &path, bool final)
  {
    std::string istring;
    std::string ostring;
    for (const auto & it : path.second) {
      istring.append(it.first);
      ostring.append(it.second);
    }
    float weight = path.first;

    if ((max_input_length > 0) &&
        (istring.length() > max_input_length))
      {
        // continue searching, break off this path
        return RetVal(true, false);
      }
    if ((max_output_length > 0) &&
        (ostring.length() > max_output_length))
      {
      return RetVal(true, false);
      // continue searching, break off this path
      }
    if (input_prefix.length() > 0)
      {
        if (istring.length() < input_prefix.length())
          {
            return RetVal(true, true);
          }
      if (istring.compare(0, input_prefix.length(), input_prefix) != 0)
        {
          return RetVal(true, false);
          // continue searching, break off this path
        }
      }
    if(output_prefix.length() > 0)
      {
        if (ostring.length() < output_prefix.length())
          {
            return RetVal(true, true);
          }
        if (ostring.compare(0, output_prefix.length(), output_prefix) != 0)
          {
            return RetVal(true, false);
            // continue searching, break off this path
          }
      }
    if (input_exclude.length() > 0 &&
        (istring.find(input_exclude) != std::string::npos))
      {
        return RetVal(true, false);
        // continue searching, break off this path
      }
    if (output_exclude.length() > 0 &&
        (ostring.find(output_exclude) != std::string::npos))
      {
        return RetVal(true, false);
        // continue searching, break off this path
      }
    if (max_weight >= 0 && weight > (max_weight + beam))
      {
        return RetVal(true, false);
        // continue searching, break off this path
      }
    // the path passed the checks. Print it if it is final
    if (final)
      {
        if (print_in_pairstring_format)
          {
        bool first_pair=true;
            for (const auto & it : path.second)
              {
        if ((not filter_fd) ||
            (not FdOperation::is_diacritic(it.first))) {
          if (print_spaces && not first_pair)
            {
              *out_ << " ";
            }

          *out_ << get_print_format(it.first);
          first_pair=false;
        }

        if (it.first.compare(it.second) != 0)
          {
            if ((not filter_fd) ||
            (not FdOperation::is_diacritic(it.second)))
              *out_ << ":"
                << get_print_format(it.second);
          }
              }
            if (display_weights)
              {
                *out_ << "\t" << path.first;
              }
            *out_ << "\n";
          }
        else
          {
        bool is_automaton=true;

        bool first_symbol=true;
            for (const auto & it : path.second)
              {
        if ((not filter_fd) ||
            (not FdOperation::is_diacritic(it.first)))
          {
            if (print_spaces && not first_symbol)
              {
            *out_ << " ";
              }
            if (it.first.compare(it.second) != 0)
              is_automaton=false;

            *out_ << get_print_format(it.first);
          }
        first_symbol=false;
          }
        if (print_spaces)
          {
        *out_ << " ";
          }

        if (not is_automaton) {
          *out_ << ":";
          for (const auto & it : path.second)
        {
          if ((not filter_fd) ||
              (not FdOperation::is_diacritic(it.second)))
            {
              if (print_spaces)
            {
              *out_ << " ";
            }
              *out_ << get_print_format(it.second);
            }
        }
        }

            if (display_weights)
              {
                *out_ << "\t" << path.first;
              }
            *out_ << std::endl;
           }
        count++;
      }
    // continue until we've printed max_num strings
    return RetVal((max_num < 1) || (count < max_num), true);
  }
};

int
process_stream(HfstInputStream& instream, std::ostream& outstream)
{
  //instream.open();

  bool first_transducer=true;
  while(instream.is_good())
  {
    if (!first_transducer && print_separator_after_each_transducer)
      outstream << "--" << std::endl;
    first_transducer=false;

    HfstTransducer t(instream);

    /* Pairstring format is not supported on optimized lookup format. */
    if (print_in_pairstring_format &&
    (instream.get_type() == HFST_OL_TYPE ||
     instream.get_type() == HFST_OLW_TYPE) ) {
      fprintf(stderr,
          "Error: option --print-in-pairstring-format not supported on "
          "       optimized lookup transducers, exiting program\n" );
      exit(1);
    }

    if(input_prefix != "")
      verbose_printf("input_prefix: '%s'\n", input_prefix.c_str());

    if(beam >= 0)
      {
        verbose_printf("Finding the weight of the best path...\n");
      try
        {
          HfstTransducer tc(t);
          tc.n_best(1);
          HfstTwoLevelPaths best_paths;
          tc.extract_paths(best_paths);
          if (best_paths.size() != 1)
            {
              error(EXIT_FAILURE, 0, "n_best(1) produced more than one path");
            }
          max_weight = best_paths.begin()->first;
        }
      catch (const FunctionNotImplementedException & e)
        {
          if (instream.get_type() == hfst::HFST_OL_TYPE ||
              instream.get_type() == hfst::HFST_OLW_TYPE)
            {
              error(EXIT_FAILURE, 0, "option --beam not implemented for optimized lookup format");
            }
          else
            {
              error(EXIT_FAILURE, 0, "option --beam not implemented");
            }
          return EXIT_FAILURE;
        }
      catch(const HfstFatalException & e)
        {
          error(EXIT_FAILURE, 0, "n_best runs out of memory");
          return EXIT_FAILURE;
        }
      }

    if(nbest_strings > 0)
    {
      verbose_printf("Pruning transducer to %i best path(s)...\n",
             nbest_strings);
      try
        {
          t.n_best(nbest_strings);
        }
      catch (const FunctionNotImplementedException & e)
        {
          if (instream.get_type() == hfst::HFST_OL_TYPE ||
              instream.get_type() == hfst::HFST_OLW_TYPE)
            {
              error(EXIT_FAILURE, 0, "option --nbest not implemented for optimized lookup format");
            }
          else
            {
              error(EXIT_FAILURE, 0, "option --nbest not implemented");
            }
          return EXIT_FAILURE;
        }
      catch(const HfstFatalException & e)
        {
          error(EXIT_FAILURE, 0, "n_best runs out of memory");
          return EXIT_FAILURE;
        }
    }
    else
    {
      if(max_random_strings <= 0 && max_strings <= 0 && max_input_length <= 0
     && max_output_length <= 0 &&
     cycles < 0 && t.is_cyclic())
      {
        error(EXIT_FAILURE, 0,
              "Transducer is cyclic. Use one or more of these options: "
              "-n, -N, -r, -l, -L, -c");
        return EXIT_FAILURE;
      }
    }

    if(max_strings > 0)
      verbose_printf("Finding at most %i path(s)...\n", max_strings);
    else if(max_random_strings > 0)
      verbose_printf("Finding at most %i random path(s)...\n",
             max_random_strings);
    else
      verbose_printf("Finding strings...\n");

    /* not random strings */
    if (max_random_strings <= 0)
      {
    Callback cb(max_strings, &outstream);
    if(eval_fd)
      t.extract_paths_fd(cb, cycles, filter_fd);
    else
      t.extract_paths(cb, cycles);
    verbose_printf("Printed %i string(s)\n", cb.count);
      }
    /* random strings */
    else
      {
    HfstTwoLevelPaths results;
    try {
      if (eval_fd) {
        t.extract_random_paths_fd(results, max_random_strings, filter_fd);
      }
      else {
        t.extract_random_paths(results, max_random_strings);
      }
    }
    catch (const FunctionNotImplementedException & e) {
      if (instream.get_type() == hfst::HFST_OL_TYPE ||
          instream.get_type() == hfst::HFST_OLW_TYPE)
        {
          error(EXIT_FAILURE, 0, "option --random not implemented for optimized lookup format");
        }
      else
        {
          error(EXIT_FAILURE, 0, "option --random not implemented");
        }
      return EXIT_FAILURE;
    }

    Callback cb(max_random_strings, &outstream);
    for (auto path : results)
      {
        cb(path, true /*final*/);
      }
    verbose_printf("Printed %i random string(s)\n", cb.count);
      }

    //if (print_separator_after_each_transducer)
    //  outstream << "--" << std::endl;
  }

  instream.close();
  return EXIT_SUCCESS;
}


int main( int argc, char **argv ) {
#ifdef WINDOWS
  _setmode(0, _O_BINARY);
#endif

  hfst_set_program_name(argv[0], "0.1", "HfstFst2Strings");
  epsilon_format = hfst_strdup("");
    int retval = parse_options(argc, argv);

    if (max_strings > 0 && max_random_strings > 0 && !silent)
      {
    fprintf(stderr,
        "warning: option --max_strings ignored, --random used\n");
    max_strings = -1;
      }

    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }
    // close buffers, we use streams
    if (inputfile != stdin)
    {
        fclose(inputfile);
    }
    if (outfile != stdout)
    {
        fclose(outfile);
    }
    verbose_printf("Reading from %s, writing to %s\n",
        inputfilename, outfilename);
    // here starts the buffer handling part
    std::unique_ptr<HfstInputStream> instream;
    try {
      instream.reset((inputfile != stdin) ?
        new HfstInputStream(inputfilename) : new HfstInputStream());
    } catch(const HfstException e) {
        fprintf(stderr, "%s is not a valid transducer file\n", inputfilename);
        return EXIT_FAILURE;
    }

    if (outfile != stdout)
    {
      std::ofstream outstream(outfilename);
      retval = process_stream(*instream, outstream);
    }
    else
      retval = process_stream(*instream, std::cout);

    free(inputfilename);
    free(outfilename);
    free(epsilon_format);
    return retval;
}

